Ubuntu setup docs (#50)
UNSWComputing/rUNSWift-docs: source/setup/nao_setup_ubuntu.rst

##################
NAO Setup (Ubuntu)
##################

Note: Now that we've moved to Ubuntu, the setup process is slightly different. Follow along to see what's changed.

Most of these instructions can be used to upgrade or reset robots as well.

The robots that return from repair will likely not need these instructions.

.. note::
 **You do not need to do this if are just setting up your environment!**

*****************
What's different?
*****************
Our robots have traditionally been running NAOqi OS. However, we recently transitioned our robots (and our tooling/software) to be compatible with Ubuntu 22.04 as a base image.

> This offers several advantages. On top of being more familiar to most of us, it makes things a lot simpler in a technical perspective,

Reviewer: ```suggestion
This offers several advantages. On top of being more familiar to most of us, it makes things a lot simpler from a technical perspective,
```

> alongside allowing us to pursue new technical projects with a better chance of success such as the migration to ROS.

Reviewer: ```suggestion
alongside allowing us to pursue new technical projects with a better chance of success such as the ongoing research into a migration to ROS/ROS2.
```

> *****************

Reviewer: There's a LOT of these. See if there's an autoformatter, a lint rule or something that just does them all for you, I haven't flagged them all as I think it's a style nitpick.

```suggestion
**************
```

Build rUNSWift

> *****************

Reviewer: ```suggestion
**************
```

> Before doing anything, you should build rUNSWift. While technically optional, this will be useful later on, as the image

Reviewer: ```suggestion
Before doing anything, you should build `runswift`. While technically optional, this will be useful later on, as the image
```

> you create would automatically bake in the rUNSWift executable if it exists. Make sure you follow the PC setup guide beforehand.

Reviewer: ```suggestion
you create would automatically bake in the ``runswift`` executable if it exists. Make sure you follow the PC setup guide beforehand.
```

> Use docker.ps1 if you're on windows.

Reviewer: ```suggestion
Use ``docker.ps1`` if you're on Windows.
```

.. code-block:: bash

    ./bin/docker.sh bin/build-runswift.sh

*****************
Download opn file
*****************

OpenNAO (opn) files are the image file for the NAO robot.

> - For teams new to RoboCup SPL, please contact the RoboCup SPL Technical Committee (rc-spl-tc@lists.robocup.org) as mentioned `here <https://spl.robocup.org/v6-support/>`_.

Reviewer: Just a suggestion, I don't know what the TC does, but I often see things like this to slow down spam bots.
```suggestion
- For teams new to RoboCup SPL, please contact the RoboCup SPL Technical Committee (rc-spl-tc AT lists.robocup.org) as mentioned `here <https://spl.robocup.org/v6-support/>`_.
```

> - If you're not aligned with RoboCup, please contact Aldebaran.

Reviewer: ```suggestion
- If you're not aligned with RoboCup, please contact your regional NAO reseller, most likely either Softbank Robotics or United Robotics Group.
```

> If you are from UNSW, running ``build-naoimage.sh`` will download the opn file for you as part of the image creation process.

Reviewer: ```suggestion
If you are on site at UNSW, running ``build-naoimage.sh`` will download the opn file for you as part of the image creation process.
```

For passwords, please :ref:`contact` us.

You can download them manually as follows:

.. code-block:: bash

    # 2.8 image (v6)

>     rsync -P repository@runswift2.cse.unsw.edu.au:/var/www/html/opennao2/build-2.8.5.1x/nao-2.8.5.11_ROBOCUP_ONLY_with_root.opn .

Reviewer: **Blocker**: I don't remember as part of Softbank legal if we can even reveal the existence of this `ROBOCUP_ONLY` file / filename. But it's now public (uh oh - GitHub support request?)

@jayenashar Do you happen to recall so we can unblock this PR at least?

Author: Pretty sure this was copied from the previous setup guide. Open to whatever is right

.. warning::
    Due to software licensing between Softbank and RoboCup SPL, teams can't publicly release the NAOqi OS provided by Softbank.

> ********************************

Reviewer: ```suggestion
****************************
```

Create a custom Ubuntu Image

> ********************************

Reviewer: ```suggestion
****************************
```

We will now create a custom Ubuntu image, which will use the base Softbank image and add our own customisations on top.

The process is based off the NaoDevil's flasher. You can find the source code `here <https://github.com/NaoDevils/NaoImage>`_.

- Windows
    .. code-block:: powershell

        bin\docker.ps1 bin/build-naoimage.sh

> - MacOS/Linux

Reviewer: ```suggestion
- macOS/Linux
```
OR
```suggestion
- Mac/Linux
```

There is no MacOS ;)

    .. code-block:: bash

        bin/docker.sh bin/build-naoimage.sh

> Ensure you have docker installed before continuing. This process will take a significant period if you are running for the first time.

Reviewer: Or `docker` if referring to the CLI instead of the app?
```suggestion
Ensure you have Docker installed before continuing. This process will take a significant period if you are running for the first time.
```

For subsequent runs, you will have the option to save time by reusing the base Ubuntu image.
As long as you didn't make any changes to the root scripts (you are unlikely to as they are located inside of the NaoDevils code) you can continue with the saved base to save time.

> *******************

Reviewer: ```suggestion
******************
```

Flashing the Robot

> *******************

Reviewer: ```suggestion
******************
```

You have 2 options to flash the robot:

Create a Flashable USB (Recommended)

> *************************************

Reviewer: ```suggestion
************************************
```

You can also use the flash script in the bin directory to flash the USB using ``dd``.
    .. code-block:: bash

        # run without args to view the help menu
        ./bin/make-usb.sh

Once the USB is made, turn off the robot and plug the USB to the back of its head. Then, hold the chest button continuously until it turns blue.

Let go, and it should start flashing blue rapidly. The lights on the side of the head will tell you the flashing progress. Once it's done, the robot will automatically boot.

.. tip::
    If this step takes too long (>30 minutes) turn the robot off and retry flashing.

Network flash
****************
You can also flash over the network with the following script:

    .. code-block:: bash

        ./bin/flash-robot.sh <robot-hostname or ip>

Restarting the robot will then begin the flashing process (the script should also prompt you to do so)

*********************************
Robot Config, Name and Wifi Setup
*********************************
**If the robots are new** you will need to:

* Add the robot to ``robots/robots.cfg``

>     * You can find its head id via ``cat /sys/qi/head_id`` after SSHing onto the robot.

Reviewer: ```suggestion
    * You can find its head id via ``cat /sys/qi/head_id`` after ``ssh <robot>``.
```

>     * Note you can flash safely without adding the robot to this file, and discover the head ID by sshing to the IP it calls out after flashing and running the ``cat`` command.

Reviewer: ```suggestion
    * Note you can flash safely without adding the robot to this file, and discover the head ID by ``ssh <IP>`` the robot calls out after flashing, and running the ``cat`` command.
```

    things roughly work as expected on the base image.

.. tip::

    ``<hostname>`` is likely ``nao.local`` for new or factory reset robots. This could be used instead of the IP address during setup.

.. tip::

    Workaround: If this last step is causing trouble for you, try changing the hostname at
    the command line first, then syncing, for example:

.. code-block:: bash

    PC$ ssh nao@nao.local
    nao$ sudo nano /etc/hostname
    # Agree, then change the hostname from 'nao' to for example 'treebeard'
    # then reboot the robot and continue as normal, i.e.
    PC$ nao_sync -s treebeard
    PC$ nao_sync -rd treebeard
    # reboot again

*********************************
Connecting to GameController Wifi
*********************************

To play a game by the SPL Rules, the ``runswift`` executable needs to send packets of
information to, and respond to commands from, one specific soccer field's
`GameController <https://github.com/RoboCup-SPL/GameController3/>`_. Typically the
competition organisers will provide the list of field SSIDs and any other details

> at the competition, for example `SPL_A` to `SPL_E` has been typical of RoboCup.

Reviewer: Double check this one, it might have already been there, or it might not have been an issue, I don't recall the exact RST syntax.

```suggestion
at the competition, for example ``SPL_A`` to ``SPL_E`` has been typical of RoboCup.
```

##################
NAO Setup (Ubuntu)
##################

Note: Now that we've moved to Ubuntu, the setup process is slightly different. Follow along to see what's changed.

Most of these instructions can be used to upgrade or reset robots as well.

The robots that return from repair will likely not need these instructions.

.. note::
 **You do not need to do this if are just setting up your environment!**

*****************
What's different?
*****************
Our robots have traditionally been running NAOqi OS. However, we recently transitioned our robots (and our tooling/software) to be compatible with Ubuntu 22.04 as a base image.

This offers several advantages. On top of being more familiar to most of us, it makes things a lot simpler in a technical perspective,
alongside allowing us to pursue new technical projects with a better chance of success such as the migration to ROS.

*****************
Build rUNSWift
*****************

> Before doing anything, you should build rUNSWIift. While technically optional, this will be useful later on, as the image

Author: Typo

you create would automatically bake in the rUNSWift executable if it exists. Make sure you follow the PC setup guide beforehand.

Use docker.ps1 if you're on windows.

.. code-block:: bash

    ./bin/docker.sh bin/build-runswift.sh

*****************
Download opn file
*****************

OpenNAO (opn) files are the image file for the NAO robot.

- For teams new to RoboCup SPL, please contact the RoboCup SPL Technical Committee (rc-spl-tc@lists.robocup.org) as mentioned `here <https://spl.robocup.org/v6-support/>`_.

- If you're not aligned with RoboCup, please contact Aldebaran.

> - If you're a member of rUNSWift, download them to your current working directory using

Author: Rephrase or remove

You can download them as follows:

.. code-block:: bash

>     # 2.1 image (v4 and v5)

Author: Don't show downloading old one, no point

The process is based off the NaoDevil's flasher. You can find the source code `here <https://github.com/NaoDevils/NaoImage>`_.

- Windows
    .. code-block:: powershell

        bin\docker.ps1 bin/build-naoimage.sh

- MacOS/Linux
    .. code-block:: bash

        bin/docker.sh bin/build-naoimage.sh

Ensure you have docker installed before continuing. This process will take a significant period if you are running for the first time.

For subsequent runs, you will have the option to save time by reusing the base Ubuntu image.
As long as you didn't make any changes to the root scripts (you are unlikely to as they are located inside of the NaoDevils code) you can continue with the saved base to save time.

*******************
Flashing the Robot
*******************

You have 2 options to flash the robot:

Create a Flashable USB (Recommended)
*************************************

This approach is likely to work with the least amount of complications.

#.  Follow the instructions on
    `Softbank's Documentation <http://doc.aldebaran.com/2-1/software/naoflasher/naoflasher.html>`_

>     .. note::

Author: A bit confusing with the below text also

You can also use the flash script in the bin directory to flash the USB.  A help menu will be displayed without args.
    .. code-block:: bash

        ./bin/make-usb.sh

Once the USB is made, turn off the robot and plug the USB to the back of its head. Then, hold the chest button continuously until it turns blue.

Let go, and it should start flashing blue rapidly. The lights on the side of the head will tell you the flashing progress. Once it's done, the robot will automatically boot.

.. tip::
    If this step takes too long (>30 minutes) turn the robot off and retry flashing.

Network flash
****************
You can also flash over the network with the following script:

    .. code-block:: bash

        ./bin/flash-robot.sh <robot-hostname or ip>

Restarting the robot will then begin the flashing process (the script should also prompt you to do so)

*********************************
Robot Config, Name and Wifi Setup
*********************************
**If the robots are new** you will need to:

* Add the robot to ``robots/robots.cfg``
    * You can find its head id via ``cat /sys/qi/head_id``.

>     * If you don't know it yet, you can flash without adding the robot to this file, and discover the head ID by sshing to the IP it calls out after flashing.

Author: And running the cat command

Uploading rUNSWift
*********************************

The image by default should already have rUNSWift. You can check details about the flash by looking at ``image.commit.sha`` and ``image.build.time`` in the home dir.

To flash a new version, run:

.. code-block:: bash

    bin/nao_sync.sh <robot hostname or ip>

The robot should already be good to go with the IP specified in robots/robots.cfg for LAN and wifi in SPL_A

You can see the network settings in ``/etc/netplan`` directory. You can modify these files and then run ``sudo netplan apply`` to apply the changes.

However, we do have scripts to manage these files as we detail further down below. Therefore, you shouldn't really need to modify them yourself.

The default username and password for the robot after flashing is ``nao:nao``.
Save yourself the hassle of typing this out repeatedly and add your key to ``image/home/nao/.ssh/authorized_keys`` and flash.

*********************************
Troubleshooting
*********************************

.. tip::

    It's sometimes useful to just flash the base opn image if you experience issues with robot kinematics, but are confident there are no hardware issues.

>     You can simply use the base image with the existing flash methods to restore the default nao image.

Author: "and then attempt to flash the custom image once you confirm things roughly work as expected in the base image"

.. tip::

    ``<hostname>`` is likely ``nao.local`` for new or factory reset robots.

    Robots that are flashed without factory reset should retain their hostnames and you don't

>     need to specify ``-h <hostname>`` and ``<robot-name>.local`` will be used automatically.

Author: Outdated

.. tip::

    Workaround: If this last step is causing trouble for you, try changing the hostname at
    the command line first, then syncing, for example:

.. code-block:: bash

    PC$ ssh nao@nao.local
    nao$ sudo nano /etc/hostname
    # Agree, then change the hostname from 'nao' to for example 'treebeard'
    # then reboot the robot and continue as normal, i.e.
    PC$ nao_sync -s treebeard
    PC$ nao_sync -rd treebeard
    # reboot again

*********************************
Connecting to GameController Wifi
*********************************

To play a game by the SPL Rules, the ``runswift`` executable needs to send packets of
information to, and respond to commands from, one specific soccer field's
`GameController <https://github.com/RoboCup-SPL/GameController3/>`_. Typically the
competition organisers will provide the list of field SSIDs and any other details
at the competition, for example `SPL_A` to `SPL_E` has been typical of RoboCup.

> One way to do this is to use the change_field script located in bin. You can also modify the WIFI network manually in the ``/etc/netplan`` directory while connected through ethernet if you're in a pinch.

Author: Not strictly true. Can be done via wifi as well
